core/commands/dag/dag.go

```diff
@@ -258,6 +273,277 @@ var DagResolveCmd = &cmds.Command{
+					failedPins++
+				}
+
+				if err := res.Emit(&CarImportOutput{Root: ret}); err != nil {
+					return err
+				}
+			}
+
+			if failedPins > 0 {
+				return fmt.Errorf(
+					"unable to pin all roots: %d out of %d failed",
+					failedPins,
+					len(roots),
+				)
+			}
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, event *CarImportOutput) error {
+
+			silent, _ := req.Options[silentOptionName].(bool)
+			if silent {
+				return nil
+			}
+
+			enc, err := cmdenv.GetLowLevelCidEncoder(req)
+			if err != nil {
+				return err
+			}
+
+			if event.Root.PinErrorMsg != "" {
+				event.Root.PinErrorMsg = fmt.Sprintf("FAILED: %s", event.Root.PinErrorMsg)
+			} else {
+				event.Root.PinErrorMsg = "success"
+			}
+
+			if !event.Root.PresentInCar {
+				event.Root.PinErrorMsg += " (root specified in .car header without its data)"
+			}
+
+			_, err = fmt.Fprintf(
+				w,
+				"Pinned root\t%s\t%s\n",
+				enc.Encode(event.Root.Cid),
+				event.Root.PinErrorMsg,
+			)
+			return err
+		}),
+	},
+}
+
+func importWorker(req *cmds.Request, re cmds.ResponseEmitter, api iface.CoreAPI, ret chan importResult) {
+
+	// this is *not* a transaction
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(req.Context, api.Dag())
+
+	roots := make(map[cid.Cid]bool)
+	var blockCount int64
+
+	it := req.Files.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+
+		// wrap a defer-closer-scope
+		//
+		// every single file in it() is already open before we start
+		// just close here sooner rather than later for neatness
+		// and to surface potential erorrs writing on closed fifos
+		// this won't/can't help with not running out of handles
+		err := func() error {
+			defer file.Close()
+
+			car, err := gocar.NewCarReader(file)
+			if err != nil {
+				return err
+			}
+
+			// Be explicit here, until the spec is finished
+			if car.Header.Version != 1 {
+				return errors.New("only car files version 1 supported at present")
+			}
+
+			for _, c := range car.Header.Roots {
+				if _, exists := roots[c]; !exists {
+					roots[c] = false
+				}
+			}
+
+			for {
+				block, err := car.Next()
+				if err != nil && err != io.EOF {
+					return err
+				} else if block == nil {
+					break
+				}
+
+				// the double-decode is suboptimal, but we need it for batching
+				nd, err := ipld.Decode(block)
+				if err != nil {
+					return err
+				}
+
+				if err := batch.Add(req.Context, nd); err != nil {
+					return err
+				}
+				blockCount++
+				if 0 == blockCount%200 {
+					// work around https://github.com/ipfs/go-ds-flatfs/issues/36 for the time being
+					// batch up-to 200 at a time ( fly under MacOS's default of 256 )
+					if err := batch.Commit(); err != nil {
```

Comment:
Did you _actually_ run into this problem? `ipfs add` has the same behavior.

Looking at the code, this is probably because we usually test with the daemon (which raises the file descriptor limit to 8196). Let's not hack a fix in here.

Reply:
Also note: we _can_ reduce the batch size when constructing the batch.

Reply:
Yes I have ran into it. Part of the sharness tests runs **without** a daemon ( as would be seen in the wild in many cases ). If you remove this block - that test will fail.

I didn't realize the batch size can be controlled, will push a fix momentarily.

Reply:
Welp the batch **size** can be controlled, and is currently [defaulting to `128`](https://github.com/ipfs/go-ipld-format/blob/v0.1.0/batch.go#L203-L206). But the batch **application** is carried out async with [`ParallelBatchCommits`](https://godoc.org/github.com/ipfs/go-ipld-format#pkg-variables) concurrent jobs. Combined together they blow through the `ulimit` since blocks come in at cid-decoding-speed.

Since this is a global variable, the only reasonable way to flip it would be to muck with `atomic.` and that doesn't sound right.

How should we fix this? Add an option-defaulting-to-0 to ignore the global value of `ParallelBatchCommits`? Not using a batch at all?

Reply:
My real concern here is that these are the default limits we're using for `ipfs add` as well. Given that, we should consider reducing the default batch size (or just fixing flatfs).

Reply:
I looked briefly at `ipfs add` and did not see other locks/governors. I suppose we are not hitting a problem with `ipfs add` sufficiently often because. 
1) it usually with a daemon running ( higher ulimit )
2) the FS contention gets in the way allowing us to fly just under the limit, by luck

Reply:
I know why...

1. We flush when we finish adding each individual file.
2. The default chunk size is 256KiB.
3. The maximum amount of data we can buffer is 8MiB * NCPU * 2.

That means we can end up with `(8196*NCPU*2)/256` outstanding blocks (usually ~256).

Regardless, we should fix the batch: https://github.com/ipfs/go-ipld-format/pull/56.

Then we should benchmark and see how this affects add performance on flatfs.

Reply:
Do we have prior art for benchmarking? A typical workstation OS with an SSD will not display a difference for obvious reasons.

Reply:
Unfortunately, no. However, I seem to remember a pretty dramatic difference when messing with these parameters last time, even on an SSD.

Reply:
That's ok, I'll improvise. Finishing up something else at the moment, will likely not look into this until way later today.

Reply:
Ok, sharness passes now with no ulimit fixes and no workarounds
Had to pull in https://github.com/ipfs/go-merkledag/pull/53 and https://github.com/ipfs/go-ipld-format/pull/55 in order to benefit from https://github.com/ipfs/go-ipld-format/pull/56 

Initial testing looks good, will have the final number tomorrow morning, day got too long